text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,882 @@
+would require a change in many unit tests across the code base with no clear
+value to users.
+
+We propose to modify the `Stack` construct such that if a stack is created
+without an explicit `scope`, an `App` instance will automatically be created and
+used as it's scope.
+
+```ts
+const stack = new Stack();
+assert(stack.node.scope instanceof App); // previously it was `undefined`
+```
+
+Since only the root construct may have an empty ID, we will also need to assign
+an ID. We propose to use `"Stack"` since we already have fallback logic that
+uses this as the stack name when the stack does not have an ID (see [stack.ts](https://github.com/aws/aws-cdk/blob/8c0142030dce359591aa76fe314f19fce9eddbe6/packages/%40aws-cdk/core/lib/stack.ts#L920)).
+
+This change will allow us to remove any special casing we have for stacks in the
+testing framework and throughout the synthesis code path (we have quite a lot of
+that), because we will be able to assume that `Stage.of(construct.node.root)` is
+never `undefined` and has a `synth()` method which returns a cloud assembly.
+
+Unit tests sometimes use "incremental tests" for synthesized templates. For
+example:
+
+```ts
+const stack = new Stack();
+const c1 = new MyConstruct(stack, 'c1', { foos: [ 'bar' ] });
+expect(stack).toHaveResource('AWS::Resource', {
+  Foos: [ 'bar' ]
+});
+
+// now add a "foo" and verify that the synthesized template contains two items
+c1.addFoo('baz');
+expect(stack).toHaveResource('AWS::Resource', {
+  Foos: [ 'bar', 'baz' ]
+});
+```
+
+Since `stage.synth()` (which is called by `expect(stack)`) would reuse the
+synthesized output if called twice, we will also need to introduce a
+`stage.synth({ force: true })` option. This will be the default behavior when
+using `expect(stack)` or `SynthUtils.synth()`.
+
+The main side effect of this change is that construct paths in unit tests will
+now change. In the above example, `foo.node.path` will change from `MyFoo` to
+`Stack/MyFoo`. Additionally, tests for resources that utilized `node.uniqueId`
+to generate names will also change given `uniqueId` is based on the path.
+
+Since app-less stacks are only used during tests, this should not have
+implications on production code, but it does break some of our test suite.
+
+### What can we do on 1.x?
+
+In order to reduce [merge conflicts](#repository-migration-efforts) between 1.x
+and 2.x we considered introducing this change on the 1.x branch prior to forking
+off 2.x.
+
+However, this is technically a breaking (behavioral) change for end-users since
+`node.path` and `node.uniqueId`, and their derivatives, will change for trees
+rooted by a `Stack`, and unit tests will need to be updated.
+
+Therefore we propose to introduce this change as a feature flag over the 1.x
+codebase and migrate all of our unit tests (I don't believe we have a way to
+enable feature flags for all unit tests, but we can devise one).
+
```

Comment:
can we mitigate this inside the `match`/`haveResourceLike`  implementations? ignoring the `Stack` prefix?

Reply:
I found a way to work around this breaking change so it's not breaking any more ("relocate").